tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx

```diff
@@ -13,44 +13,96 @@
+        fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::string &treeName, const std::vector<std::string> &fileNames, REntryRange entryRange = {})
+      : fTreeNames(std::vector<std::string>{treeName}), fFileNameGlobs(fileNames), fStartEntry(entryRange.fStartEntry),
+        fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                REntryRange entryRange = {})
+      : fTreeNames(
+           fileNames.size() != treeNames.size() && treeNames.size() != 1
+              ? throw std::logic_error("RDatasetSpec exepcts either N trees and N files, or 1 tree and N files.")
+              : treeNames),
+        fFileNameGlobs(fileNames), fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   void AddFriend(const std::string &treeName, const std::string &fileName)
+   {
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
+      fFriendInfo.fFriendFileNames.emplace_back(std::vector<std::string>{fileName});
+      fFriendInfo.fFriendChainSubNames.emplace_back(std::vector<std::string>{}); // this is a tree
+   }
 
-   RDatasetSpec(const std::string &datasetName, const std::string &fileName, ULong64_t startEntry = 0,
-                ULong64_t endEntry = 0, const std::vector<std::string> &defaultColumns = {},
-                const std::vector<std::string> &subTreenames = {})
-      : fDatasetName(datasetName), fFileNameGlobs(std::vector<std::string>{fileName}), fStartEntry(startEntry),
-        fEndEntry(endEntry), fDefaultColumns(defaultColumns), fSubTreeNames(subTreenames)
+   void AddFriend(const std::string &treeName, const std::vector<std::string> &fileNames)
    {
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
+      fFriendInfo.fFriendFileNames.emplace_back(fileNames);
+      fFriendInfo.fFriendChainSubNames.emplace_back(std::vector<std::string>{treeName}); // now this is a chain
    }
 
-   RDatasetSpec(const std::string &datasetName, const std::vector<std::string> &fileNames, ULong64_t startEntry = 0,
-                ULong64_t endEntry = 0, const std::vector<std::string> &defaultColumns = {},
-                const std::vector<std::string> &subTreenames = {})
-      : fDatasetName(datasetName), fFileNameGlobs(fileNames), fStartEntry(startEntry), fEndEntry(endEntry),
-        fDefaultColumns(defaultColumns), fSubTreeNames(subTreenames)
+   void AddFriend(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames)
    {
+      if (fileNames.size() != treeNames.size() && treeNames.size() != 1)
+         throw std::logic_error("RDatasetSpec's friend exepcts either N trees and N files, or 1 tree and N files.");
```

Comment:
```suggestion
      if (fileNames.size() != treeNames.size())
         throw std::logic_error("The input treeNames and fileNames vectors must be of the same size.");
```

Reply:
Alright, a small note is that this gives "less" freedom to the user. And for us handling this does not come with extra cost. I have no idea of the potential use case, so I am going to apply the change.